[DOC] api conversion tool (#4492)
strimzi/strimzi-kafka-operator: documentation/modules/upgrading/con-upgrade-custom-resources.adoc

// This module is included in the following assemblies:
//
// assembly-upgrade-resources.adoc

[id='con-upgrade-custom-resources-{context}']

= API versioning

[role="_abstract"]

> Custom resources are created as instances of APIs added by CRDs to extend Kubernetes resources.

Reviewer: This is just a suggestion...

```suggestion
Custom resources are manipulated using APIs added to Kubernetes by CRDs. 
Put another way, CRDs extend the Kubernetes API to allow creation of custom resources.
```

Author: Edited and controlled sounds good: _Custom resources are edited and controlled using APIs added to Kubernetes by CRDs._

Author: > Custom resources are edited and controlled using APIs added to Kubernetes by CRDs.
> Put another way, CRDs extend the Kubernetes API to allow creation of custom resources.

> CRDs are installed in a Kubernetes cluster to define the schemas used to instantiate and manage the custom resources.

Reviewer: Again, just a suggestion:

```suggestion
CRDs are themselves resources within Kubernetes. 
They are installed in a Kubernetes cluster to define the versions of API for the custom resource. Each version of the CR API can define its own schema for that versions.
```

> For release {ProductVersion}, Strimzi CRDs must specify `{ApiVersion}` as the API version and `{KafkaApiVersion}` as the API path for custom resources.

Reviewer 2: I think this might confuse the timing. The new API version is really required only for Strimzi 0.23 - the next minor version.

> For release {ProductVersion}, Strimzi CRDs must specify `{ApiVersion}` as the API version and `{KafkaApiVersion}` as the API path for custom resources.

Reviewer 2: ```suggestion
For release {ProductVersion}, Strimzi CRDs must specify `{ApiVersion}`.
```

Reviewer: What does "Strimzi CRDs must specify `{ApiVersion}`" actually mean though? Is this a statement about the CRDs or the CRs? Remember that a CRD has lots of versions in it, so it's ambigous and you need to be specific.

Reviewer 2: Honestly, when I look at it now I'm not sure what it is trying to say.

Author: How about if we change to this:

> For release {ProductVersion}, Strimzi custom resources must specify `{ApiVersion}` as the API version.
> Strimzi Operators access the custom resources served by the Kubernetes API server using a URL path (_API path_) which includes the API version.

Reviewer 2: I'm still not completely sure what would be the take from this. It would make sense to say that _From release 0.23 you have to use v1beta2_. The 0.22 actually uses v1beta1 for most cases (but not all). It just adds v1beta2 to the CRDs etc.

Author: With the changes to the opening paragraph, maybe we can drop the first line and bring the second line up:

![image](https://user-images.githubusercontent.com/47596553/110140619-2f949880-7dcc-11eb-865e-0c04e2f7d89c.png)

Reviewer 2: Yeah, I think this is OK for me.

> Strimzi Operators access custom resources using the API path.

Reviewer: ```suggestion
Strimzi Operators access custom resources of the Kubernetes API server using a URL path ("API path") which includes the CR version.
```

Reviewer 3: Nice explanation.

The introduction of `{ApiVersion}` updates the schemas of the custom resources.
Older API versions are deprecated.

The `v1alpha1` API version is deprecated for the following Strimzi resources:

* `Kafka`
* `KafkaConnect`
* `KafkaConnectS2I`
* `KafkaConnector`
* `KafkaMirrorMaker`
* `KafkaMirrorMaker2`
* `KafkaTopic`
* `KafkaUser`
* `KafkaBridge`
* `KafkaRebalance`

The `v1beta1` API version is deprecated for the following Strimzi resources:

* `Kafka`
* `KafkaConnect`
* `KafkaConnectS2I`
* `KafkaMirrorMaker`
* `KafkaTopic`
* `KafkaUser`

> 

Reviewer 2: We should add something like

```
IMPORTANT: The `v1alpha1` and `v1beta1` versions will be removed in the next minor release.
```

> NOTE: Deprecated API versions might also be simultaneously supported by a CRD, as long as the schema for the custom resources is valid.

Reviewer 2: I would remove this, I think it might cause more confusion then explain things.

API versioning policy indicates the relative stability of the version.
Alpha version APIs are the least mature, beta versions more mature, and stable versions represent a durable release.

Example API versions:

> * Alpha version: `v1alpha1`
> * Beta version: `v1beta1`
> * Stable version: `v1`

Reviewer 2: Maybe adding `v1beta2` which we actually use now might help .

```suggestion
* Alpha version: `v1alpha1`
* Beta version: `v1beta1`
* Another beta version: `v1beta2`
* Stable version: `v1`
```

Author: I added on the same line:

* Alpha version: `v1alpha1`
* Beta version: `v1beta1`, `v1beta2`
* Stable version: `v1`

After a new custom resource type is added to a cluster by installing a CRD, you can create instances of the resource based on its specification.

> Strimzi CRDs strictly adhere to a structural schema specification.

Reviewer: Actually it's only v1beta2 which is structural.

Author: Changed to: _Custom resources using `{ApiVersion}` strictly adhere to a structural schema specification._

Structural schemas are based on the OpenAPI v3.0 validation schema.
OpenAPI v3.0 provides validation for the creation of custom resources.

A complete structural schema:

* Follows OpenAPI structure and syntax standards
* Specifies the types for _all_ properties
* Allows only intended values for properties

> For example, the specification for a Kafka topic specifies that a topic customer resource must have at least one partition and one replica to be valid.

Reviewer 3: ```suggestion
For example, the specification for a Kafka topic specifies that a topic custom resource must have at least one partition and one replica to be valid.
```

Author: :facepalm:

.Example CRD for topic custom resources
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion} <1>
kind: CustomResourceDefinition
metadata:
  name: kafkatopics.kafka.strimzi.io
  #...
spec:
  group: kafka.strimzi.io <2>
  #...
  versions: <3>
  - name: {ApiVersion}
    served: true
    storage: true
    #...
  validation:
    openAPIV3Schema: <4>
      properties:
        spec:
          type: object
          properties:
            partitions:
              type: integer
              minimum: 1
            replicas:
              type: integer
              minimum: 1
              maximum: 32767
      #...
----
<1> Specifies the API path for custom resources using the API group and API version.
<2> The API group (domain) name.
<3> The supported API versions. The API version declared as the _storage_ version is used to create new custom resource instances.

> <4> Schema validation used when creating custom resources.

Reviewer: TBH I don't think that Strimzi users really need to know this level of detail, do they?

Author: Maybe. It's to give a bit of context to OpenAPI and structual schema. It was part of the spec when we first discussed the content needed.

Author: removed

// This module is included in the following assemblies:
//
// assembly-upgrade-resources.adoc

[id='con-upgrade-custom-resources-{context}']

= API versioning

[role="_abstract"]
Custom resources are edited and controlled using APIs added to Kubernetes by CRDs.
Put another way, CRDs extend the Kubernetes API to allow creation of custom resources.
CRDs are themselves resources within Kubernetes.
They are installed in a Kubernetes cluster to define the versions of API for the custom resource.
Each version of the custom resource API can define its own schema for that version.

> For release {ProductVersion}, Strimzi custom resources must specify `{ApiVersion}` as the API version.

Reviewer: I'm not sure what this is supposed to mean. You've not explained what happens when a client uses a URL with version X and the resource is stored at version Y (!= X). For our CRDs kube will work fine it you access a v1beta2 resource though a v1beta1 URL. I think the point we're trying to make is:

1. We want people to use create `v1beta2` resources (and the procedures will cover upgrading their existing resources to `v1beta2`)
2. If people have API server clients they should be using the `v1beta2` URLs.

Author: The line is removed now. Do we want to expand on this still?

Author: ![image](https://user-images.githubusercontent.com/47596553/110146341-5950be00-7dd2-11eb-962c-74725237d67b.png)

Reviewer: Last sentence:

Kubernetes clients (including the Strimzi operators) access the customer resources...

> Strimzi Operators access the custom resources served by the Kubernetes API server using a URL path (_API path_) which includes the API version.

Reviewer: I think this would be better on line 15, and it's not just the Strimzi operators, but all clients.

Author: Ah, just mentioned that!

> Strimzi Operators access the custom resources served by the Kubernetes API server using a URL path (_API path_) which includes the API version.

Reviewer 2: I would remove this and replace it with just something like _You should use v1beta2 from now on._

Author: This is outdated. See screen capture on https://github.com/strimzi/strimzi-kafka-operator/pull/4492#discussion_r588419934 to see if we need further chnage

* Follows OpenAPI structure and syntax standards
* Specifies the types for _all_ properties
* Allows only intended values for properties

For example, the specification for a Kafka topic specifies that a topic custom resource must have at least one partition and one replica to be valid.

.Example CRD for topic custom resources
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion} <1>
kind: CustomResourceDefinition
metadata:
  name: kafkatopics.kafka.strimzi.io
  #...
spec:
  group: kafka.strimzi.io <2>
  #...
  versions: <3>
  - name: {ApiVersion}
    served: true
    storage: true
    #...
  validation:
    openAPIV3Schema: <4>
      properties:
        spec:
          type: object
          properties:
            partitions:
              type: integer
              minimum: 1
            replicas:
              type: integer
              minimum: 1
              maximum: 32767
      #...
----
<1> Specifies the API path for custom resources using the API group and API version.
<2> The API group (domain) name.
<3> The supported API versions. The API version declared as the _storage_ version is used to create new custom resource instances.

> <4> Schema validation used when creating custom resources.

Reviewer: Honestly, I'd like to see this deleted unless we can point to a concrete reason a user needs to know this. I don't think they do.

Author: Removed
